modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/metrics/LongAsyncCounterAdapter.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.telemetry.apm.internal.metrics;
+
+import io.opentelemetry.api.metrics.Meter;
+import io.opentelemetry.api.metrics.ObservableLongCounter;
+
+import org.elasticsearch.common.util.concurrent.ReleasableLock;
+import org.elasticsearch.telemetry.apm.AbstractInstrument;
+import org.elasticsearch.telemetry.metric.LongAsyncCounter;
+import org.elasticsearch.telemetry.metric.LongWithAttributes;
+
+import java.util.Objects;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Supplier;
+
+public class LongAsyncCounterAdapter extends AbstractInstrument<ObservableLongCounter> implements LongAsyncCounter {
+    private final Supplier<LongWithAttributes> observer;
+    private final ReleasableLock closedLock = new ReleasableLock(new ReentrantLock());
+    private boolean closed = false;
+
+    public LongAsyncCounterAdapter(Meter meter, String name, String description, String unit, Supplier<LongWithAttributes> observer) {
+        super(meter, name, description, unit);
+        this.observer = observer;
+    }
+
+    @Override
+    protected ObservableLongCounter buildInstrument(Meter meter) {
+        var builder = Objects.requireNonNull(meter).counterBuilder(getName());
+        return builder.setDescription(getDescription()).setUnit(getUnit()).buildWithCallback(measurement -> {
+            LongWithAttributes observation;
+            try {
+                observation = observer.get();
+            } catch (RuntimeException err) {
+                assert false : "observer must not throw [" + err.getMessage() + "]";
+                return;
+            }
+            measurement.record(observation.value(), OtelHelper.fromMap(observation.attributes()));
+        });
+    }
+
+    @Override
+    public void close() throws Exception {
+        try (ReleasableLock lock = closedLock.acquire()) {
```

Comment:
Sorry, late to the game @pgomulka. Just curious, do you really need to block when closing here? I'm also wondering, I see a lot of places where `ReleasableLock` (using a `ReentrantLock`)  is used for rather simple use cases that could be handles using an already existing monitor. Is that just a convention in the code base?

Reply:
I am not sure what would be the behaviour of instrument being closed multiple times (not documented). Here using a lock makes sure that clsoed flag is atomically set when instrument is closed.
Regarding the ReleasableLock I would call it a convention. It was just convenient to use here.

Reply:
Actually, unfortunately, this doesn't work reliably. When another thread reads `closed` it might read an outdated value and run `getInstrument().close()` again.

Here's an alternative using CAS without locking:
```Java
    private AtomicBoolean closed = new AtomicBoolean(false);

    @Override
    public void close() throws Exception {
        if(closed.compareAndSet(false, true)){
            getInstrument().close();
        }
    }
```    
Or alternatively, if locking is considered necessary:
```
    private volatile boolean closed = false;

    @Override
    public void close() throws Exception {
        synchronized (observer) {
            if (closed == false) {
                getInstrument().close();
                closed = true;
            }
        }
    }
```

Reply:
good ideas. I think the approach with closed is slightely difference as it would set the flag to true even if the method failed (although not sure if this is possible)
the second approach is pretty much the same as the one with ReleasableLock.

Reply:
Regarding CAS on closed, I think on that level that's what you want. Have a look at the recommendations in the Java docs of `Closeable` / `AutoCloseable` :

> Cases where the close operation may fail require careful attention by implementers. It is strongly advised to relinquish the underlying resources and to internally mark the resource as closed, prior to throwing the exception.